### ISSUE TITLE
Command / option to assign optimally sized refreservation after original refereservation has been deleted

### ISSUE THREAD

**Issue by Binarus** [openzfs/zfs#11399]
### Describe the feature would like to see added to OpenZFS

When we create a ZVOL via `zfs create` and do not give `-o refreservation=...` at the command line, ZFS automatically calculates an optimal value for `refreservation` and sets it on the new ZVOL. Furthermore, it re-calculates and sets the optimal value each time the ZVOL's size changes.

However, when we remove `refreservation` from the ZVOL, or tamper with it and set it too low (making the ZVOL a sparse volume), there is no way to set it to the optimal value again, i.e. to the value ZFS would set it to when creating the ZVOL.

This is a problem because (unless we are willing to figure it out from the source code) we have no clue how to calculate the optimal value. Obviously, `refreservation` having too low means that the ZVOL can run out of space, causing unexpected situations and putting data at risk, and having it too high means a waste of space (think of snapshots).

That calculating the optimal size is not straightforward is shown by the following fact: When creating a new ZVOL of `100 GB` in size with `-b 512`, `refreservation` is set to `150 GB`. When creating a new ZVOL of the same size with `-b 4096`, `refreservation` is set to something below `110 GB` (I don't have the exact number right now).

Neither we nor anybody else we have talked to could even explain which other properties (besides `volblocksize`) ZFS takes into account when calculating `refreservation`, let alone the exact way of the calculation.

Therefore, we believe that it should be possible to let ZFS automatically calculate and add `refreservation` to a ZVOL at any time, even, and notably, if there is no `refreservation` at all on that ZVOL, or that ZVOL is sparse.

### How will this feature improve OpenZFS?

Deleting `refreservation` from a ZVOL is currently one-way. Obviously, the minimal reasonable value of `refreservation` cannot be computed easily from `volsize`. On the other hand, giving it a higher value would be a waste (think of snapshots).

There are many reasons for removing `refreservation` from a ZVOL, or to set it to a small value, thereby making the ZVOL sparse (accidentally, temporarily because we have run out of space, and so on). Currently, there is no way to re-add `refreservation` with the optimal value.

We believe that it would be a noticeable improvement to ZFS if there was a command to add `refreservation` with the optimal value at any time to a ZVOL. Of course, the optimal value is the value ZFS would automatically set if it would newly create the ZVOL.

We further believe that it would be easy to provide such a command, because the respective code already is in ZFS: When newly creating a ZVOL, `refreservation` is automatically set to the optimal value (if you don't prevent it by command line options).

### Additional context

No additional context.

**Comment by Binarus** [comment]
I just saw by accident that we already have this feature! It is just not documented.

I am extremely worried why nobody has closed this request because it's already implemented. It seems that nobody has read this request yet. Really?

Closing it myself ...

**Comment by MichaelHierweck** [comment]
How can the default refreservation be restored?

**Comment by behlendorf** [comment]
It'd be great to get a PR open which updates the documentation accordingly.

**Comment by Binarus** [comment]
@MichaelHierweck 

    set refreservation=auto rpool/zvol

Yes, it's that easy, but it seems to be barely know even amongst the developers. I saw it here by accident while actually searching for something completely different:

https://smartos.org/bugview/OS-6767

That issue report made me just try it in OpenZFS 2.0.0, and there it worked. Please note that I did not try it with other ZFS versions. If you could test with other version and report the result, that would be great.

I am also not sure whether this method will restore **all** behavior of automatically created `refreservation`s. For example, according to the manual, ZFS will automatically adjust `refreservation` when `volsize` is changed, but (according to the manual) **only when `refreservation` has not been tampered with in a way that the ZVOL has become sparse**.

@behlendorf 

Before documenting it: Do you know whether `set refreservation=auto` to an existing ZVOL is *exactly* the same as having the `refreservation` created automatically when creating the ZVOL? Notably, will the `refreservation` always be automatically adjusted in the same way when `volsize` changes?

Thank you very much, and best regards,

Binarus

**Comment by Binarus** [comment]
Update / @behlendorf 

I had also initiated a discussion about this subject on the mailing list. It turned out that `refreservation=auto` is not documented in previous versions, which is the reason why I didn't see it, although it works there, at least with 0.8.6 (and probably more, but nobody has tested yet). But `refreservation=auto` is documented in the newest version 2.0.0; I guess it happened when restructuring the man pages.

For that discussion, see [here](https://zfsonlinux.topicbox.com/groups/zfs-discuss/T8ed7aa2f2bcdbb33/how-to-correctly-calculate-or-set-the-refreservation-a-zvol-needs).

So there is neither need to implement that feature (because it already is) nor need for a PR to update the documentation (because `man zfsprops` for OpenZFS 2.0.0 already has it, and making a PR for older versions wouldn't make sense, would it?).

**Comment by behlendorf** [comment]
@Binarus thanks for following up.  It does seems like everything is in order.  Here's the relevant git commit, e8b2196b1464d60fd47b8606dfa46236f5d8ccb0.  It looks like this functionality was added to the 0.8.x series and the documentation was updated accordingly.   I'm sure it was easy to overlook in that huge man page!  It's nice to have them split up now.

**Comment by Binarus** [comment]
@behlendorf Thank you very much again!

I never have seen the man page for 0.8.x, because debian buster (my preferred distro) still comes with 0.7.12, where it hasn't been implemented yet. I got my hands on an OpenZFS 2.0.0 system to test it after I had seen the link I have posted above. I wasn't given much time on that system, so couldn't study the man pages for 2.0.0 there.

It is clearly my fault that I didn't mention my ZFS version in the first post. Sorry for the fuss about nothing ... Still wondering why Google did not show any reasonable results when I was researching the issue; probably my fault either (not being a native English speaker).

Thanks again, and best regards!
